Bottom: 3b721cca9626318d51ebc6127f94f8e3c1de7633
Top:    33a1f211f7f32674efb946acdda1590a0302fc9d
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-08-28 21:26:39 +0800

Refresh of vhost_net-tx-batching

---

diff --git a/drivers/vhost/vhost.c b/drivers/vhost/vhost.c
index e15f581..2058f8d 100644
--- a/drivers/vhost/vhost.c
+++ b/drivers/vhost/vhost.c
@@ -2586,8 +2586,7 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 	int ret = 0;
 	u16 last_avail_idx, last_used_idx, total;
 	__virtio16 avail_idx, *idx = indices;
-	int start;
-	struct vring_used_elem heads[64], *head;
+	struct vring_used_elem heads[64];
 	struct vring_used_elem __user *used;
 	int i;
 
@@ -2626,7 +2625,6 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 
 	total = ret;
 	last_used_idx = vq->last_used_idx;
-	head = &heads[0];
 	while (total) {
 		int ret2;
 		u16 copied = vq->num - (last_used_idx & (vq->num - 1));
@@ -2634,7 +2632,7 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 		copied = min(copied, total);
 		ret2 = vhost_copy_to_user(vq,
 				&vq->used->ring[last_used_idx & (vq->num - 1)],
-				head, copied * sizeof *used);
+				&heads[ret - total], copied * sizeof *used);
 
 		if (unlikely(ret2)) {
 			vq_err(vq, "Failed to update used ring!\n");
@@ -2642,7 +2640,6 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 		}
 
 		total -= copied;
-		head += copied;
 		last_used_idx += copied;
 	}
